### PR TITLE
More `not_null<>` improvements, some general improvements

### PR DESCRIPTION
**Pull request by mbeutel**
* Add macros `gsl_noreturn`, `gsl_nodiscard`, `gsl_deprecated()` (https://github.com/martinmoene/gsl-lite/commit/66f62f8ddf77e218771aa2fc0e195e922c46a144)

* CMake package now installs only <gsl/gsl-lite.hpp> by default (https://github.com/martinmoene/gsl-lite/commit/ff9bd3411adac51040db0b8b82863b0c22104c39)

* Fix implicit conversion operators and operators `->` and `*` of `not_null<>` (https://github.com/martinmoene/gsl-lite/commit/4c060db5100c873020142ecb8d11305c499f033f, https://github.com/martinmoene/gsl-lite/commit/dcc811e1c2829706152d39fce3dc802abcfd2b0b, https://github.com/martinmoene/gsl-lite/commit/9122aa558e3bfdc57269787ceab1cd8e34f12c6b, https://github.com/martinmoene/gsl-lite/commit/ae5ea3e136060c5aaa5d88214019cb2186aa9066)

* `not_null<>` comparison operators are now transparent and use SFINAE (https://github.com/martinmoene/gsl-lite/commit/c67b5373090781528b49531195aa08e23d33641c)

* Add configuration option `gsl_CONFIG_NOT_NULL_TRANSPARENT_GET` (https://github.com/martinmoene/gsl-lite/commit/a84560ef71816e4b44d5d8a18b34fbf071e23ee6)

* Add deduction guides for `not_null<>` (https://github.com/martinmoene/gsl-lite/commit/7a6278305c8ff644736c5d798bc5591dc15e5051)

* Add `make_not_null()` (https://github.com/martinmoene/gsl-lite/commit/042d4749891ccac2deca844d9d331b44c7b47e2d)

...and many minor fixes (https://github.com/martinmoene/gsl-lite/commit/c23820dab29bad198b909c754cd6ad6e9f3807ec, https://github.com/martinmoene/gsl-lite/commit/bb6532a75f1959b657e56855cef87d7ad21810bb, https://github.com/martinmoene/gsl-lite/commit/867ad12fb0489dafe34ef03f05cae06d460d8f7b, https://github.com/martinmoene/gsl-lite/commit/a8409320d0ea8451b70cbb84eae2cfb6a257b59e) and improvements (https://github.com/martinmoene/gsl-lite/commit/44812ef990342c999621f3e547e80f74125c5d36, https://github.com/martinmoene/gsl-lite/commit/75c5d15fe47ff6018bffd5450153e35b50210fe7, https://github.com/martinmoene/gsl-lite/commit/b81d806aeddc9d4b0c475241f92135e35b283b8b, https://github.com/martinmoene/gsl-lite/pull/202/commits/19a880e9ccf85424cb2f90bf980c683189c39dd6, https://github.com/martinmoene/gsl-lite/pull/202/commits/bdace076649fcb9170b442a073e5b4d19952ea73, https://github.com/martinmoene/gsl-lite/pull/202/commits/55769e3524138b2c7bce09ac17e08585dbf8d032).

### CODE REVIEW

File: include/gsl/gsl-lite.hpp [hunk 1]
```diff
@@ -1484,21 +1544,32 @@ class not_null
...
-    gsl_api gsl_constexpr   operator T const &     () const & { return checked_ptr(); }
-    gsl_api gsl_constexpr14 operator T &&          () &&      { return std::move(checked_ptr()); }
+    gsl_api gsl_constexpr14 operator T() const & { return checked_ptr(); }
+    gsl_api gsl_constexpr14 operator T() &&      { return std::move(checked_ptr()); }
+
+    gsl_api gsl_constexpr14 T const & operator->() const & { return checked_ptr(); }
+    gsl_api gsl_constexpr14 T &&      operator->() &&      { return std::move(checked_ptr()); }
+
+    gsl_api gsl_constexpr14 element_type &  operator*() const & { return *checked_ptr(); }
+    gsl_api gsl_constexpr14 element_type && operator*() &&      { return *std::move(checked_ptr()); }
```
Comment (petamas) [inline include/gsl/gsl-lite.hpp]: 1. Why do we need separate const& and && versions of these operators? Is there any use case when this matters?
2. You changed the return type of operator-> in the "default" (non-constref-get, non-transparent-get) build. Was that your intention?

**Comment by mbeutel** [comment]
> I messed up and have written a few comments before I figured out how I can send a review. Sorry for the confusion!

No problem. Thanks for reviewing! :)


> My main problem is that the relation of gsl_CONFIG_NOT_NULL_GET_BY_CONST_REF and gsl_CONFIG_NOT_NULL_TRANSPARENT_GET is not apparent from neither the documentation, nor the code. Can they be both set to 1? If so, what happens? If not, why don't we have an error? (Like with the contract violation configs)

The docs state that `gsl_CONFIG_NOT_NULL_GET_BY_CONST_REF` affects the return value of `get()` "and friends". The latter is the reason why the options can be combined; even with `gsl_CONFIG_NOT_NULL_TRANSPARENT_GET=1`, `gsl_CONFIG_NOT_NULL_GET_BY_CONST_REF` still affects the conversion operator (if function ref qualifiers are not available) and `operator->()`.

I'll try to clarify this.

File: include/gsl/gsl-lite.hpp (hunk 1 above)
Comment (mbeutel) [reply include/gsl/gsl-lite.hpp]: Cf. https://github.com/martinmoene/gsl-lite/commit/a84560ef71816e4b44d5d8a18b34fbf071e23ee6#commitcomment-35452682

File: test/not_null.t.cpp [hunk 2]
```diff
@@ -128,23 +148,23 @@ CASE( "not_null<>: Allows to construct a const pointer from a non-null underlyin
 
     not_null< const int* > p( &i );
 
-    EXPECT( p.get() == &i );
+    EXPECT( p == &i );
```
[the remark this answers is not in the document]
Comment (mbeutel) [reply test/not_null.t.cpp]: Which implicit conversions are you referring to?

In any case, `not_null<>` should be comparable to anything the underlying pointer is comparable with.

File: test/not_null.t.cpp [hunk 3]
```diff
@@ -119,7 +119,27 @@ CASE( "not_null<>: Allows to construct from a non-null underlying pointer (raw p
 
     not_null< int* > p( &i );
 
-    EXPECT( p.get() == &i );
+    EXPECT( p == &i );
+}
+
+CASE( "not_null<>: Allows to construct from a non-null underlying pointer (raw pointer) with make_not_null()" )
```
[the remark this answers is not in the document]
Comment (mbeutel) [reply test/not_null.t.cpp]: cv qualifiers should be trivially lost during inference, and `make_not_null()` is agnostic regarding the constness of pointees, so I'm not sure I see a benefit in this.

File: test/not_null.t.cpp [hunk 4]
```diff
@@ -403,6 +443,42 @@ CASE( "not_null<>: Allows to construct from a non-null underlying pointer (uniqu
     EXPECT( &*p == raw );
 }
 
+CASE( "not_null<>: Allows to move from a not-null pointer (unique_ptr)" )
```
[the remark this answers is not in the document]
Comment (mbeutel) [reply test/not_null.t.cpp]: Agreed.

**Comment by mbeutel** [comment]
> I'll try to clarify this.

I think `operator ->()` can safely return `T const &` in every case. That would leave the return value of `get()` (unless `gsl_CONFIG_NOT_NULL_TRANSPARENT_GET=1`) and the conversion operator (unless function ref qualifiers are available).

At this point, I'd like to reconsider the option of removing the option entirely, with both the pre-C++11 conversion operator and the legacy `get()` returning `T const &` unconditionally too. Could this possibly break any code?

File: test/not_null.t.cpp (hunk 3 above)
[the remark this answers is not in the document]
Comment (petamas) [reply test/not_null.t.cpp]: I'll try to figure out what my example was, and get back to you :)

**Comment by petamas** [comment]
> At this point, I'd like to reconsider the option of removing the option `gsl_CONFIG_NOT_NULL_GET_BY_CONST_REF` entirely, with both the pre-C++11 conversion operator and the legacy `get()` returning `T const &` unconditionally too. Could this possibly break any code?

I think for 0.x, we should leave the return values as they were in the "default" case, then in 1.0, remove  `gsl_CONFIG_NOT_NULL_GET_BY_CONST_REF`, make `gsl_CONFIG_NOT_NULL_TRANSPARENT_GET=1` the default, and if it's 0, then the code should behave as if `gsl_CONFIG_NOT_NULL_GET_BY_CONST_REF=1` was defined. This way, we could drop lots of ifdefs in 1.0, have the most modern version as the default, but could  fall back to a more-or-less compatible older solution. What do you think?

**Comment by petamas** [comment]
> Please do. `remove_cv<>` shouldn't be needed unless there is a compiler bug we need to work around.

I could not reproduce the problem; I think it might have happened when I was using U&&, and I just misremembered it persisting with U. So consider this resolved. :)

(Is there a way I can reply inline? I can't find a button I could use around your comment...)

**Comment by mbeutel** [comment]
> I think for 0.x, we should leave the return values as they were in the "default" case, then in 1.0, remove `gsl_CONFIG_NOT_NULL_GET_BY_CONST_REF`, make `gsl_CONFIG_NOT_NULL_TRANSPARENT_GET=1` the default, and if it's 0, then the code should behave as if `gsl_CONFIG_NOT_NULL_GET_BY_CONST_REF=1` was defined. This way, we could drop lots of ifdefs in 1.0, have the most modern version as the default, but could fall back to a more-or-less compatible older solution. What do you think?

That is what I will do if I cannot convince myself that unconditional `gsl_CONFIG_NOT_NULL_GET_BY_CONST_REF=1` cannot possibly break code.

**Pull request merged**